File: clients/reactive-dart/lib/src/contracts.dart
import 'dart:convert';

/// Declared latency/durability strategy for one application write.
enum SyncOptimism { remoteConfirmed, localDurable, localThenRemote }

enum SyncSource {
  local,
  http,
  websocket,
  tcp,
  supabase,
  broadcast,
  serviceWorker,
}

enum SyncRecordAuthority { localView, authoritative }

/// Structural mirror of shared-auth's verified identity.
///
/// Bearer and refresh tokens are deliberately absent so storage keys, stream
/// dedupe identities, and diagnostics cannot leak credentials.
final class SyncSessionIdentity {
  const SyncSessionIdentity({
    required this.sharedUserId,
    required this.provider,
    required this.providerTenant,
    required this.providerSubject,
    this.project,
    this.supabaseUserId,
    this.sessionId,
    this.authority,
    this.roles = const <String>[],
  });

  factory SyncSessionIdentity.fromSharedAuthJson(Map<String, Object?> json) {
    String requiredString(String key) {
      final value = json[key];
      if (value is! String || value.isEmpty) {
        throw FormatException('shared-auth identity requires $key');
      }
      return value;
    }

    final roles = json['roles'];
    return SyncSessionIdentity(
      sharedUserId: requiredString('shared_user_id'),
      provider: requiredString('provider'),
      providerTenant: requiredString('provider_tenant'),
      providerSubject: requiredString('provider_subject'),
      project: json['project'] as String?,
      supabaseUserId: json['supabase_user_id'] as String?,
      sessionId: json['session_id'] as String?,
      authority: json['authority'] as String?,
      roles: roles is List<Object?>
          ? List<String>.unmodifiable(roles.whereType<String>())
          : const <String>[],
    );
  }

  final String sharedUserId;
  final String provider;
  final String providerTenant;
  final String providerSubject;
  final String? project;
  final String? supabaseUserId;
  final String? sessionId;
  final String? authority;
  final List<String> roles;
}

sealed class SyncSession {
  const SyncSession();
}

final class AuthenticatedSyncSession extends SyncSession {
  const AuthenticatedSyncSession(this.identity);
  final SyncSessionIdentity identity;
}

final class AnonymousSyncSession extends SyncSession {
  const AnonymousSyncSession();
}

final class UnauthenticatedSyncSession extends SyncSession {
  const UnauthenticatedSyncSession();
}

final class DegradedSyncSession extends SyncSession {
  const DegradedSyncSession(this.reason);
  final String reason;
}

SyncSessionIdentity requireAuthenticated(SyncSession session) {
  if (session case AuthenticatedSyncSession(:final identity)) return identity;
  if (session case DegradedSyncSession(:final reason)) {
    throw StateError(
      'sync authentication is degraded; privileged synchronization fails closed: $reason',
    );
  }
  throw StateError(
      'sync requires an authenticated session: ${session.runtimeType}');
}

String _component(String value) => Uri.encodeComponent(value);

/// Durable storage/queue partition. Session rotation does not strand work.
String storagePartitionKey(SyncSessionIdentity identity) => <String>[
      identity.provider,
      identity.providerTenant,
      identity.sharedUserId,
    ].map(_component).join(':');

/// Live transport generation. Session rotation cancels stale streams.
String transportSessionKey(SyncSessionIdentity identity) =>
    '${storagePartitionKey(identity)}:${_component(identity.sessionId ?? 'sessionless')}';

final class SyncRecordEvent<T> {
  const SyncRecordEvent({
    required this.table,
    required this.recordId,
    required this.operation,
    required this.payload,
    required this.revision,
    required this.source,
    required this.authority,
    required this.sessionPartition,
    this.checkpoint,
    this.pending = false,
    this.dedupeKey,
  });

  final String table;
  final String recordId;
  final String operation;
  final T? payload;
  final String revision;
  final String? checkpoint;
  final SyncSource source;
  final SyncRecordAuthority authority;
  final bool pending;
  final String? dedupeKey;
  final String sessionPartition;
}

Object? _stable(Object? value) {
  if (value is List<Object?>) return value.map(_stable).toList(growable: false);
  if (value is Map<Object?, Object?>) {
    final entries = value.entries
        .map((entry) => MapEntry(entry.key.toString(), _stable(entry.value)))
        .toList(growable: false)
      ..sort((left, right) => left.key.compareTo(right.key));
    return Map<String, Object?>.fromEntries(entries);
  }
  return value;
}

String stableJson(Object? value) => jsonEncode(_stable(value));

String recordEventDedupeKey<T>(SyncRecordEvent<T> event) =>
    event.dedupeKey ??
    <String>[
      event.sessionPartition,
      event.table,
      event.recordId,
      event.operation,
      event.revision,
      stableJson(event.payload),
    ].join('\u0000');

File: clients/reactive-ts/src/contracts.ts
export const SYNC_OPTIMISM = Object.freeze({
  remoteConfirmed: 'remote-confirmed',
  localDurable: 'local-durable',
  localThenRemote: 'local-then-remote',
} as const);

export type SyncOptimism =
  (typeof SYNC_OPTIMISM)[keyof typeof SYNC_OPTIMISM];

export type SyncSource =
  | 'local'
  | 'http'
  | 'websocket'
  | 'tcp'
  | 'supabase'
  | 'broadcast'
  | 'service-worker';

/**
 * Structural subset of shared-auth's verified identity contract.
 *
 * The package intentionally does not import the private shared-auth package at
 * runtime. A Supabase session or a generated shared-auth binding can map into
 * this shape without leaking bearer tokens into IndexedDB, BroadcastChannel,
 * logs, or event dedupe keys.
 */
export interface SyncSessionIdentity {
  shared_user_id: string;
  provider: string;
  provider_tenant: string;
  provider_subject: string;
  project?: string | null;
  supabase_user_id?: string | null;
  session_id?: string | null;
  authority?: 'shared-auth' | 'supabase' | string;
  roles?: readonly string[];
}

export type SyncSession =
  | {
      status: 'authenticated';
      identity: SyncSessionIdentity;
    }
  | { status: 'anonymous' }
  | { status: 'unauthenticated' }
  | { status: 'degraded'; reason: string };

/** Stable queue/storage partition. Token refresh and session rotation keep it. */
export function storagePartitionKey(identity: SyncSessionIdentity): string {
  return [identity.provider, identity.provider_tenant, identity.shared_user_id]
    .map((part) => encodeURIComponent(part))
    .join(':');
}

/** Transport generation. A new revocation-aware session cancels stale streams. */
export function transportSessionKey(identity: SyncSessionIdentity): string {
  return `${storagePartitionKey(identity)}:${encodeURIComponent(
    identity.session_id ?? 'sessionless',
  )}`;
}

export function requireAuthenticated(
  session: SyncSession,
): SyncSessionIdentity {
  if (session.status === 'authenticated') return session.identity;
  if (session.status === 'degraded') {
    throw new Error(
      `sync authentication is degraded; privileged synchronization fails closed: ${session.reason}`,
    );
  }
  throw new Error(`sync requires an authenticated session (${session.status})`);
}

export type SyncRecordOperation = 'upsert' | 'delete';
export type SyncRecordAuthority = 'local-view' | 'authoritative';

export interface SyncRecordEvent<T> {
  table: string;
  recordId: string;
  operation: SyncRecordOperation;
  payload: T | null;
  /** Server revision or local mutation identity, represented losslessly. */
  revision: string;
  checkpoint?: string;
  source: SyncSource;
  authority: SyncRecordAuthority;
  /** True while the local projection still includes unacknowledged work. */
  pending?: boolean;
  /** Cross-transport identity. Must not include access/refresh tokens. */
  dedupeKey?: string;
  sessionPartition: string;
}

export interface SyncHint {
  reason:
    | 'local-mutation'
    | 'remote-change'
    | 'connectivity'
    | 'background-wake'
    | 'manual';
  source: SyncSource;
  sessionPartition: string;
  table?: string;
  recordId?: string;
  checkpoint?: string;
}

function stable(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(stable);
  if (value && typeof value === 'object') {
    return Object.fromEntries(
      Object.entries(value as Record<string, unknown>)
        .sort(([left], [right]) => left.localeCompare(right))
        .map(([key, child]) => [key, stable(child)]),
    );
  }
  return value;
}

export function stableJson(value: unknown): string {
  return JSON.stringify(stable(value));
}

export function recordEventDedupeKey<T>(event: SyncRecordEvent<T>): string {
  return (
    event.dedupeKey ??
    [
      event.sessionPartition,
      event.table,
      event.recordId,
      event.operation,
      event.revision,
      stableJson(event.payload),
    ].join('\u0000')
  );
}

export function sameProjectedValue<T>(left: T | null, right: T | null): boolean {
  return stableJson(left) === stableJson(right);
}
